src/swarm-tui/Swarm/TUI/Controller.hs

```diff
@@ -692,46 +701,47 @@ handleREPLEventTyping m = \case
           -- Otherwise, just move around in the history as normal.
           _ -> modify $ adjReplHistIndex Newer
       ControlChar 'r' ->
-        Brick.zoom (uiGameplay . uiREPL) $ do
+        Brick.zoom (scenarioState . uiGameplay . uiREPL) $ do
           uir <- get
           let uinput = uir ^. replPromptText
           case uir ^. replPromptType of
             CmdPrompt _ -> replPromptType .= SearchPrompt (uir ^. replHistory)
             SearchPrompt rh -> forM_ (lastEntry uinput rh) $ \found ->
               replPromptType .= SearchPrompt (removeEntry found rh)
-      CharKey '\t' -> do
+      CharKey '\t' -> Brick.zoom scenarioState $ do
         s <- get
         let names = s ^.. gameState . baseEnv . envTypes . to assocs . traverse . _1
         uiGameplay . uiREPL %= tabComplete (CompletionContext (s ^. gameState . creativeMode)) names (s ^. gameState . landscape . terrainAndEntities . entityMap)
         modify validateREPLForm
-      EscapeKey -> do
+      EscapeKey -> Brick.zoom scenarioState $ do
         formSt <- use $ uiGameplay . uiREPL . replPromptType
         case formSt of
           CmdPrompt {} -> continueWithoutRedraw
           SearchPrompt _ -> resetREPL "" (CmdPrompt [])
       ControlChar 'd' -> do
-        text <- use $ uiGameplay . uiREPL . replPromptText
+        text <- use $ scenarioState . uiGameplay . uiREPL . replPromptText
         if text == T.empty
-          then toggleModal QuitModal m
+          then toggleEndScenarioModal QuitModal m
           else continueWithoutRedraw
```

Comment:
This is the only keyboard input that requires the wider `PlayState`, rather than the narrow `ScenarioState`.